### ISSUE TITLE
Crash after clearing all shaders

### ISSUE THREAD

**Issue by mcabel** [libretro/RetroArch#8974]
## Description

If one clears all shaders either through the menu options "clear all passes", or manually through the menu remove, one by one, the the program will crash.

### Expected behavior

It should be possible to clear all shaders, to have none.

### Actual behavior

[issue 1, frequent]The application crashes. Sometimes the shader autosaves, and if it was a troublesome shader, [issue 2, rare] the application won't boot until shaders are turned off in the cfg file.

### Steps to reproduce the bug

During gametime, press f5, go to the shaders menu.

### Bisect Results

[Try to bisect and tell us when this started happening]. Load a shader preset, such as xbr or smaa or anything. Check that the shader worked in the game. Press f5 again and go to shader menu. Clear all passes(or remove passes one b one). In this way, the app crashes most often.

Alternatively, do the same but load 3-4 passes. Then clear them, wither one by one or with clear all passes. In this way, it crashes less often.

### Version/Commit
You can find this information under Information/System Information

- RetroArch: [version/commit]

### Environment information
Tested with vulcan mostly. I think also with opengl

![image](https://user-images.githubusercontent.com/10392829/59557796-fb05fb00-8fb0-11e9-82ba-f2bcd3d915b0.png)

Version shown on the screenshot
![image](https://user-images.githubusercontent.com/10392829/59557777-a2cef900-8fb0-11e9-9595-6ecde6af51fd.png)

- OS: Windows 8.1

**Comment by LazyBumHorse** [comment]
Uh oh, did I do this?
Spoilers: Yes, I did!

Can reproduce. Happens with vulkan but seemingly not with gl, doesn't matter if quick menu or desktop menu is used.
The only message between going from lots of shaders to 0 shaders and crashing (segfaulting) is this:
> [INFO] Saved shader preset to RetroArch\shaders\retroarch.slangp.

Works before https://github.com/libretro/RetroArch/commit/b54a360ab8b584a5350a125c00e605960a26297b (https://github.com/libretro/RetroArch/pull/8946).

Let's see what's going wrong...

**Comment by LazyBumHorse** [comment]
shader_vulkan.cpp:
`shader->pass[shader->passes - 1]` without any check.

**Comment by LazyBumHorse** [comment]
And it reaches that because `video_shader_read_conf_cgp` unlike before sets the shader path and thus vulkan thinks the 0 pass preset is a "proper shader" and assumes it has at least 1 pass.
Hm... wouldn't be the worst idea to set the path to `NULL`, okay, pull request soon.

**Comment by mcabel** [comment]
Also, I found this:

This might be or might not be the same problem, but I'll mention it since it seems pretty similar:
on f5, on the view/ settings menu/ video if you choose a filter, browse and use one. Then go ingame and you change your mind, there is not an option to "clear". So you "can clear" it by deleting the path on the filter text box.

if you do, it's probably not set up to accept a null value. So it can crash here too.

![image](https://user-images.githubusercontent.com/10392829/59568629-5d133e80-904b-11e9-83b7-12e01b6d1837.png)

**Comment by LazyBumHorse** [comment]
So, good news and bad news:
Good news, fixing the menu crash is easy and done.
Bad news, RetroArch currently automatically saves a default preset ("retroarch.slangp") and it'll probably also crash the next time you load it up, which is a big deal.
More bad news: Since glcore uses much the same logic than vulkan, it also has this problem.

One could patch in a safety check, but that requires (another) file read in both `retroarch_get_shader_preset` _and_ `video_driver_set_shader` (for video driver initialization and manually setting shader respectively), it'd be more efficient to fix the video drivers.

For now it's probably best to quickly revert my problematic commit, until there's a good fix for the root cause.

**Comment by LazyBumHorse** [comment]
@mcabel That second issue you're describing might be caused by the video restart that happens when you change the filter, it probably loads the left-behind retroarch.slangp preset I mentioned, try deleting.

Hm... setting and clearing a filter is very glitchy for me (in 1.7.7 at least), it just restarted the video driver about a dozen times, how odd.
I think this happens when enter is pressed.

I can't reproduce that particular crash after revert, so looks like we are good on that.

Btw, I'd recommend using shaders only if you can, because filters run on the CPU and are thus slow.

**Comment by mcabel** [comment]
@LazyBumHorse  [Yeah, super slow I can see. I was just tweaking around to discover what does what]
-Also, I tried deleting retroarch.slang, then loaded a filter. Fine until here.
-Then I deleted the textbox, to stop using the filter. This worked partially [it did not stopped using the filter, but did not crash]. It's definitively trying to load something. Anyway, Next time I went to the filter menu, It goes into an infinite loop.

- I don't know retroarch's logic internally, but this is a possible solution I can imagine.:
+when first run, retroarch could create a null filter and a null shader/shader preset. (which is internally void/null and does no graphic modification, perhaps only initializing/resetting graphics devices).
+These files should be read only.
+The system should assign nullFilter and nullShaderPreset by default at first run.
+ "clear all shaders" or remove one pass, if no other shader is available, it will set the nullShader. So it will always have some shader.
+With the filters I would add a button to clear filters, Which would set the nullFilter. Also deleting the textbox then pressing enter should set nullFilter ( OR instead, the filter textbox can be grey/readOnly, and may only be modified with "browse" and with a clear/reset filter button )

**Comment by LazyBumHorse** [comment]
@mcabel Hm... video filters don't work at all for me in Beetle HW PSX, but they do in Snes9x and isn't Beetle what you're using in this test as well?

> Then I deleted the textbox, to stop using the filter (...) it did not stopped using the filter

I cannot reproduce this, if the field is deleted and it loses focus, e.g. you click somewhere else in the options menu, it stops using the filter.

Stupid question, but are you sure it's actually the filter you are seeing?
Like try Darken.filt and clear it, is the screen still much darker?

If the filter really does not clear, can you test what it says in the controller menu under Settings (gear icon) -> Video -> Video Filter (right at the bottom)?
You can press space to clear/reset it there, which should work.

> Next time I went to the filter menu, It goes into an infinite loop

Can't really reproduce that either, but there seems to be an issue with how the desktop menu handles the focus loss event (?) of the video filter text box.
Sometimes it kind of gets in a long loop, maybe because the main window is put into foreground when it reloads and thus the focus changes, which causes another focus loss event.
Clicking into another field (so that it has focus, i.e. has a blue border) should stop the loop after a certain while.

**Comment by LazyBumHorse** [comment]
@mcabel Ah, btw, the problematic commit is reverted and it looks like it's in [nightly on the buildbot](https://buildbot.libretro.com/nightly/windows/x86_64/RetroArch.7z) already, so you can upgrade to fix the shader crashes.

**Comment by mcabel** [comment]
@LazyBumHorse I updated the application, and the issue with the shaders is gone.
So we can probably assume the thing with the filters(i.e. darken.filt ) is likely a totally different issue. So, anyhow, the shader thing is solved.

I'm not planning to use filters, but how do I "clear" the filter if I don't wanna use anything. My last configuration saved the last filter I used.

Anyhow, even if this is something else, I'll leave a link to a video showing the filter loop, for reference.
https://filebin.net/ze6weeb0byc3cc5s?fbclid=IwAR2dKEOndTuUw7YOCmbf-gkRlHJ3QaxBHH8l1UuFeo17HH0Mi2hgQSm40PE

**Comment by LazyBumHorse** [comment]
@mcabel Thanks for the video, yeah that loop is what I'm seeing as well, I'll create an issue thread for that.

For now you can use the controller menu to set and clear video filters, quoting myself:
> in the controller menu under Settings (gear icon) -> Video -> Video Filter (right at the bottom)
You can press space to clear/reset it there, which should [always] work.

But one thing still confuses me: I made a screenshot after every filter you tried and to me, the images look completely identical, especially darken.filt should be immediately obvious, but in the screenshots I don't see any effect except slow framerate (and from what I can tell from the surroundings, the camera didn't seem to autocorrect).

Are you _really really_ sure the filter do anything in Beetle HW PSX?
Just as a demonstration, this is how extreme no filter vs darken.filt is (in Snes9x):
![normal](https://user-images.githubusercontent.com/31176526/59666332-eabe6d80-91b4-11e9-8db4-bb9c68bd6c81.png)
![darken](https://user-images.githubusercontent.com/31176526/59666333-eb570400-91b4-11e9-941a-67107a78a014.png)

If you can confirm that filters don't work in Beetle HW PSX (they don't for me), we can create another issue thread for that as well.

**Comment by mcabel** [comment]
@LazyBumHorse  I was under the impression that filters worked, but seems I was mistaken. Texture filter from another menu reverted to the last saved configuration, and somehow I was fooled into thinking xbr  filter worked when it didn't.

BUT after some testing, I found that the filter is not totally "not working", but rather it is not being rendered. Because I can feel just a light drop in fps with "darken", and heavy drop in fps with "xbr".

On that same menu, maybe there's a checkbox called "bilinear filter" which I'm not 100% sure if it's not working or if the effect is so subtle(well I guess it should be subtle). If it applies filters in the same way than the .filt files, it might be part of the same issue.

before bilinear checkbox
![image](https://user-images.githubusercontent.com/10392829/59713336-31c05900-91dd-11e9-8138-6959a07d1061.png)
After bilinear checkbox checked
![image](https://user-images.githubusercontent.com/10392829/59713380-5288ae80-91dd-11e9-8dfe-310668673edd.png)

**Comment by LazyBumHorse** [comment]
> BUT after some testing, I found that the filter is not totally "not working", but rather it is not being rendered. Because I can feel just a light drop in fps with "darken", and heavy drop in fps with "xbr".

Yeah, I have no idea how those filters work, but they seem to hog your CPU whether they actually do work or not, so if they don't work, just disable them.

> On that same menu, maybe there's a checkbox called "bilinear filter" which I'm not 100% sure if it's not working or if the effect is so subtle

You probably always want to leave bilinear filtering on, it's about how the core output is upscaled to your window/screen.
Turning if off uses nearest neighbor scaling (popular for pixel art) which is only any good if it scales in integer multiples (2x, 3x...) and on 1080p, you basically never get exact integer scaling, resulting in wonky pixels.
There's an "Integer Scale" option which, well, forces integer scaling, but you'll have black bars.

If you want super sharp pixels, the retro/sharp-bilinear shader is probably the best option, it does integer mutliple nearest neighbor followed by bilinear scaling so you get the best of both worlds.

> If that the case, then nothing can be done probably, aside from redesign the filters to support other color schemes ... Which sounds, unnecessary, considering shaders have better performance and look better

Yeah, that's why they aren't recommended, they are basically only useful for systems that don't have shader support, a last resort, so to speak.

So I guess there's no problem left then, if so you can close this issue.

**Comment by LazyBumHorse** [comment]
@mcabel Ah, one last quick tip: if you're using 3D games in Beetle HW PSX, you may want to try upping the "Internal GPU resolution" under Quick Menu -> Options (once a game is loaded).

**Comment by mcabel** [comment]
@LazyBumHorse Thanks for the tip. I have it on 4x, which is as much as my laptop gpu can handle(and look good enough for me). I ended up here by fine-tuning something to make the 2d interface look better. What I found was core options/texture filtering/xbr....but it's a bit heavy for my machine. There's sabr, but I don't love how the text looks... I was trying to fin a compromise by using sabr with maybe some shader on top..... but then somebody on the forum ported a bumpmap shader to slang for me so, I'll probably go with something towards that direction...
https://forums.libretro.com/t/bump-map-or-emboss-vulcan-shader-slang/22860

Anyway, it was cool meeting you and taking through this bug. Thanks and best of luck :)
